public/app/features/variables/adhoc/actions.ts

```diff
@@ -0,0 +1,164 @@
+  operator: string;
+}
+
+const filterTableName = 'Filters';
+
+export const applyFilterFromTable = (options: AdHocTableOptions): ThunkResult<void> => {
+  return async (dispatch, getState) => {
+    let variable = getVariableByOptions(options, getState());
+
+    if (!variable) {
+      dispatch(createAdHocVariable(options));
+      variable = getVariableByOptions(options, getState());
+    }
+
+    const index = variable.filters.findIndex(f => f.key === options.key && f.value === options.value);
+
+    if (index === -1) {
+      const { value, key, operator } = options;
+      const filter = { value, key, operator, condition: '' };
+      return await dispatch(addFilter(variable.uuid!, filter));
+    }
+
+    const filter = { ...variable.filters[index], operator: options.operator };
+    return await dispatch(changeFilter(variable.uuid!, { index, filter }));
+  };
+};
+
+export const changeFilter = (uuid: string, update: AdHocVariabelFilterUpdate): ThunkResult<void> => {
+  return async (dispatch, getState) => {
+    const variable = getVariable(uuid, getState());
+    dispatch(filterUpdated(toVariablePayload(variable, update)));
+    await dispatch(variableUpdated(toVariableIdentifier(variable), true));
+  };
+};
+
+export const removeFilter = (uuid: string, index: number): ThunkResult<void> => {
+  return async (dispatch, getState) => {
+    const variable = getVariable(uuid, getState());
+    dispatch(filterRemoved(toVariablePayload(variable, index)));
+    await dispatch(variableUpdated(toVariableIdentifier(variable), true));
+  };
+};
+
+export const addFilter = (uuid: string, filter: AdHocVariableFilter): ThunkResult<void> => {
+  return async (dispatch, getState) => {
+    const variable = getVariable(uuid, getState());
+    dispatch(filterAdded(toVariablePayload(variable, filter)));
+    await dispatch(variableUpdated(toVariableIdentifier(variable), true));
+  };
+};
+
+export const setFiltersFromUrl = (uuid: string, filters: AdHocVariableFilter[]): ThunkResult<void> => {
+  return async (dispatch, getState) => {
+    const variable = getVariable(uuid, getState());
+    dispatch(filtersRestored(toVariablePayload(variable, filters)));
+    await dispatch(variableUpdated(toVariableIdentifier(variable), true));
+  };
+};
+
+export const changeVariableDatasource = (datasource: string): ThunkResult<void> => {
+  return async (dispatch, getState) => {
+    const { editor } = getState().templating;
+    const variable = getVariable(editor.id, getState());
+
+    const loadingText = 'Adhoc filters are applied automatically to all queries that target this datasource';
+
+    dispatch(
+      changeVariableEditorExtended({
+        propName: 'infoText',
+        propValue: loadingText,
+      })
+    );
+    dispatch(changeVariableProp(toVariablePayload(variable, { propName: 'datasource', propValue: datasource })));
+
+    const ds = await getDatasourceSrv().get(datasource);
+    console.log('ds', ds);
+
+    if (!ds || !ds.getTagKeys) {
```

Comment:
Just out of curiosity, we have this action just to show that message right?

Reply:
Yes, but it also sets the datasource that you have selected.